// getComicSalesDate/saveSalesInfo.js
/**************************************************************************************************
 * モジュール名　: 新刊情報登録処理
 * 概要　　　　　: 取得した新刊情報をDBに登録する
 *************************************************************************************************/

// モジュール読み込み
var Q = require('q'),
    connDb = require('cmn/connDb');

/*
 * 新刊情報登録処理
 * 概　要 : 取得した新刊情報をDBに登録する
 * 引　数 : in_comicList 取得した新刊情報リスト
 * 　　　 : in_db データベースオブジェクト
 * 戻り値 : promise（resolve：なし）
 */
function saveComicList(in_comicList, in_db){

  var _d = Q.defer();

  Q.when(in_db)
    // データベースオープン
    .then(connDb.openDb)
    .fail(_d.reject)
    // コミックマスタ情報登録
    .then(function(in_conn){
      return upsertComicList(in_comicList, in_conn);
    })
    .fail(_d.reject)
    // データベースをクローズして次の処理へ
    .done(function(){
      in_db.close();
      _d.resolve();
    });

  return _d.promise;
}

/*
 * 新刊情報リスト登録処理
 * 概　要 : 新刊情報リストをデータベースに登録する。
 * 引　数 : in_comicList 取得した新刊情報リスト
 * 　　　 : in_conn コネクション
 * 戻り値 : promise（resolve：なし）
 */
function upsertComicList(in_comicList, in_conn){

  var _d = Q.defer();

  // リストが取得出来ている場合のみ登録
  if(in_comicList.length > 0){

    // コレクションを指定
    var collection = in_conn.collection('salesinfos');
    var _args = {
      idx : 0,
      comicList : in_comicList,
      collection : collection
    };
    // リストをデータベースに登録する
    var _q = Q.when(_args);
    for(var _i=0,_len=in_comicList.length;_i<_len;_i++){
      _q = _q.then(upsertComic);
    }

    _q.done(_d.resolve);

  }else{
    // リスト情報が無い場合はログを出力して次の処理へ
    logger.warn('No List');
    _d.resolve();
  }
  return _d.promise;
}

/*
 * 新刊情報登録処理
 * 概　要 : １件分の新刊情報をデータベースに登録する。
 * 引　数 : in_args
 * 戻り値 : promise（resolve：受け取った引数）
 */
function upsertComic(in_args){

  var _d = Q.defer();

  // コミックリストから１件分の情報を取得
  var _comic = in_args.comicList[in_args.idx];

  // コミック情報をupsertする
  in_args.collection.update({
    title:_comic.title},
    _comic,
    {upsert:true},
    function(in_err){
      if(in_err){
        logger.warn(in_err.message);
      }
      // 正常に登録できたら次の処理へ
      in_args.idx++;
      _d.resolve(in_args);
    }
  );
  return _d.promise;
}

module.exports = saveComicList;
